lib/ip_blocklist_critbit.js

```diff
@@ -0,0 +1,428 @@
+ *    if (check("1.2.3.4")) {
+ *      console.log("it was in the set!")
+ *    }
+ *
+ * They're split up like this so that you can do the preparation
+ * ahead of time, and also to allow the matching function to be
+ * as efficient as possible.
+ *
+ * We represent a set of IPv4 CIDRs as an array of 32-bit integers
+ * storing both the values themselvs, and an index through which to
+ * search on them.  The idea is to keep all of the data densely packed
+ * in memory, so while things will work using an ordinary javascript
+ * array, you  might like to use an ArrayBuffer instead.
+ *
+ * The structure of the data is [metadata][index][values], as
+ * follows:
+ *
+ *  - The first metadata item, data[0], gives the offset of the
+ *    end of the index, i.e. the location of the first value in
+ *    in the set.
+ *
+ *  - The second metadata item, data[1], gives the initial bitmask
+ *    to use when traversing the index.  This bitmask selects the
+ *    the first bit at which the values in the set differ from one
+ *    another.
+ *
+ *  - The index items, data[2] through data[data[0]-1], are nodes
+ *    in a radix-tree style index based on djb's "crit-bit tree"
+ *    data structure [1].
+ *
+ *  - The value items, data[data[0]] onwards, are individual IPs or
+ *    CIDRs to be matched against.  The item at data[data[0]] is
+ *    always a sentinel value of zero, indicating no match.
+ *
+ * A lookup in this structure takes as input a 32-bit integer
+ * representing an IPv4 address.  It traverses the index by
+ * looking at the bits of this integer in order, until it finds
+ * a reference to one of the IP values in the set (if it finds zero
+ * then the input value is not in the set).
+ *
+ * The index represents a radix tree as a densely-packed array
+ * of 32-bit integers.  Each integer is a tree node encoding data
+ * according to the following rules:
+ *
+ *  - If the high bit is 0 then it's a reference
+ *    to a value item.  If the high bit is 1 then
+ *    it's a reference to an internal tree node.
+ *
+ *  - For value items:
+ *
+ *      - the bottom 5 bits give 1 minus the length of the
+ *        value's CIDR prefix.  In other words, the number of
+ *        low-order bits to ignore when comparing the input value
+ *        to the found value (from 0 to 31).
+ *
+ *      - the remaining 26 bits give the offset of the value,
+ *        from the end of the index data.
+ *
+ *      - the value 0 acts like a null pointer and represents
+ *        a failure to find a data item.
+ *
+ *  - For internal tree nodes:
+ *
+ *      - the bottom 5 bits give the number of bits to shift
+ *        the bitmask by for the next lookup (1 through 31).
+ *    
+ *      - the remaining 26 bits give the offset of the next
+ *        index node to look at, from the start of the data.
+ *
+ *  - When examining input bit b at an internal node
+ *    with offset n, we look at data[n] when input[b]
+ *    is zero and data[n+1] when input[b] is one.
```

Comment:
I read this sectio a few times but still had trouble getting my head around it, a diagram would help greatly.